src/PluginsManager/Microsoft.Performance.Toolkit.PluginsManager.Core/AvailablePlugin.cs

```diff
@@ -2,62 +2,97 @@
+        ///     The <see cref="IPluginFetcher"/> this plugin uses to fetch plugin content.
+        /// </param>
         public AvailablePlugin(
-            PluginIdentity pluginIdentity,
-            PluginSource pluginSource,
-            string displayName,
-            string description,
-            Uri packageUri,
-            Guid fetcherTypeId) 
+            AvailablePluginInfo pluginInfo,
+            IPluginDiscoverer discoverer,
+            IPluginFetcher fetcher)
         {
-            this.Identity = pluginIdentity;
-            this.PluginSource = pluginSource;
-            this.DisplayName = displayName;
-            this.Description = description;
-            this.PluginPackageUri = packageUri;
-            this.FetcherTypeId = fetcherTypeId;
+            this.AvailablePluginInfo = pluginInfo;
+            this.pluginDiscoverer = discoverer;
+            this.pluginFetcher = fetcher;
         }
 
         /// <summary>
-        ///     Gets the identity of this plugin.
-        /// </summary>
-        public PluginIdentity Identity { get; }
-        
-        /// <summary>
-        ///     Gets the souce where this plugin is discovered.
+        ///     Gets the <see cref="AvailablePluginInfo"/> associated with this plugin.
         /// </summary>
-        public PluginSource PluginSource { get; }
+        public AvailablePluginInfo AvailablePluginInfo { get; }
 
-        /// <summary>
-        ///     Gets the human-readable name of this plugin.
-        /// </summary>
-        public string DisplayName { get; }
 
         /// <summary>
-        ///     Gets the user friendly description of this plugin.
-        /// </summary>   
-        public string Description { get; }
+        ///     Gets the metadata of this plugin.
+        /// </summary>
+        /// <param name="cancellationToken">
+        ///     Signals that the caller wishes to cancel the operation.
+        /// </param>
+        /// <returns>
+        ///     The metadata of a plugin.
+        /// </returns>
+        public Task<PluginMetadata> GetPluginMetadata(CancellationToken cancellationToken)
+        {
+            return this.pluginDiscoverer.GetPluginMetadataAsync(this.AvailablePluginInfo.Identity, cancellationToken);
+        }
 
         /// <summary>
-        ///     Gets the URI where the plugin package can be fetched.
+        ///     Gets all available versions of this plugin.
         /// </summary>
-        public Uri PluginPackageUri { get; }
+        /// <param name="cancellationToken">
+        ///     Signals that the caller wishes to cancel the operation.
+        /// </param>
+        /// <returns>
+        ///      A collection of available plugins.
+        /// </returns>
+        public async Task<IReadOnlyCollection<AvailablePlugin>> GetAllVersions(CancellationToken cancellationToken)
+        {
+            IReadOnlyCollection<AvailablePluginInfo> pluginInfos = await this.pluginDiscoverer.DiscoverAllVersionsOfPlugin(
```

Comment:
There's a slight catch here... suppose we have two discoverer feeds, A and B
A has versions 1.0 and 2.0
B has versions 2.0 and 3.0

When we fetched the latest version, we returned B's version of 3.0

if we get all versions, we'll only see 2.0 and 3.0. Ideally we would get 1.0, 2.0 and 3.0

Reply:
Would it therefore make sense to keep this method on the plugins manager? maybe passing in a PluginIdentity instead of the AvailablePlugin?